effective_hand_strength.py

```diff
@@ -0,0 +1,101 @@
+from pypokerengine.engine.card import Card
+from pypokerengine.engine.hand_evaluator import HandEvaluator
+from utils import get_all_possible_cards, get_all_combinations
+
+BEHIND = 0
+AHEAD = 1
+TIE = 2
+
+
+class HandStrengthEval:
+    """
+    Implements the Effective Hand Strength as described by https://en.wikipedia.org/wiki/Poker_Effective_Hand_Strength_(EHS)_algorithm
+    """
+    def __init__(self):
+        """
+        Compute all possible hands once before hand.
+        """
+        self.all_possible_hands = get_all_combinations(get_all_possible_cards(), 2)
+        self.hand_evaluator = HandEvaluator()
+
+    def rank(self, our_hand, board_cards):
+        """
+        A simple wrapper around the hand_evaluator in pypoker engine. Used to convert strings to cards
+        :param our_hand: A list of strings containing the 2 cards in our hand
+        :param board_cards: A list of strings containing 0 to 5 cards on the board
+        :return: rank as integer
+        """
+        hand_deck = list(map(Card.from_str, our_hand))
+        cards = list(map(Card.from_str, board_cards))
+        return self.hand_evaluator.eval_hand(hand_deck, cards)
+
+    def hand_strength(self, our_hand, board_cards):
+        """
+        Evaluates the hand strength i.e P(winning| cards in hand and cards on board) + 0.5*P(tie| cards in hand
+        and cards on board). Executes pretty fast (within real time)
+        :param our_hand: The cards in our hand. Must be a list of 2 strings
+        :param board_cards: The cards on the board. Can be list of length 0 to 5 of strings.
+        :return: hand strength as float
+        """
+        ahead = tied = behind = 0
+
+        our_rank = self.rank(our_hand, board_cards)
+        for hand in self.all_possible_hands:
+            if hand[0] in our_hand + board_cards or hand[1] in our_hand + board_cards:
```

Comment:
What is hand[0] and hand[1] specifically? If I traced the code correctly, is `h[0]="SA"` and `h[1]="DA"`? If so, why does this `continue`?

Reply:
Yes your trace is correct. But continue in this case means that if we have seen the card before, don't consider them. This means we dont need to calculate combinations from scratch again which is expensive.

Reply:
👍  Thanks! Leaving this conversation open so the others can read.